File: erc8004-agents/search-service/src/indexer/reputation.ts
/** Reputation indexer — fetch and store reputation scores. */

import { ethers } from "ethers";
import type { ChainConfig } from "../utils/chains.js";
import type { AgentDatabase } from "../storage/database.js";

const REPUTATION_ABI = [
  "function getAggregateScore(uint256 agentId) view returns (uint256 score, uint256 totalReviews)",
  "event ScoreSubmitted(uint256 indexed agentId, string domain, uint256 score, address indexed reviewer)",
];

/**
 * Fetch and update reputation scores for agents on a chain.
 */
export async function updateReputationScores(
  db: AgentDatabase,
  chain: ChainConfig,
  agentIds: number[]
): Promise<void> {
  if (!chain.reputationRegistry || agentIds.length === 0) return;

  try {
    const provider = new ethers.JsonRpcProvider(chain.rpcUrl);
    const contract = new ethers.Contract(
      chain.reputationRegistry,
      REPUTATION_ABI,
      provider
    );

    for (const tokenId of agentIds) {
      try {
        const [score, count] = await contract.getAggregateScore(tokenId);
        const agent = db.getAgent(chain.chainId, tokenId);
        if (agent) {
          agent.reputationScore = Number(score);
          agent.reputationCount = Number(count);
          agent.updatedAt = new Date().toISOString();
          db.upsertAgent(agent);
        }
      } catch {
        // Agent may not have reputation data yet
      }
    }
  } catch (err) {
    console.error(`[reputation] Error updating scores on ${chain.name}:`, err);
  }
}

File: erc8004-agents/sdks/rust/src/reputation.rs
//! ReputationRegistry interactions.

use crate::types::ReputationScore;
use crate::{Error, Result};

/// Helper for ReputationRegistry operations.
pub struct ReputationRegistry {
    /// Contract address
    pub address: String,
    /// Chain ID
    pub chain_id: u64,
}

impl ReputationRegistry {
    /// Create a new ReputationRegistry helper.
    pub fn new(address: impl Into<String>, chain_id: u64) -> Self {
        Self {
            address: address.into(),
            chain_id,
        }
    }

    /// Submit a reputation score for an agent.
    pub async fn submit_score(
        &self,
        _agent_id: u64,
        _domain: &str,
        _score: u64,
        _evidence: &str,
    ) -> Result<String> {
        Err(Error::ContractError("Requires alloy provider setup".into()))
    }

    /// Get an agent's reputation score for a specific domain.
    pub async fn get_score(&self, _agent_id: u64, _domain: &str) -> Result<ReputationScore> {
        Err(Error::ContractError("Requires alloy provider setup".into()))
    }

    /// Get an agent's aggregate reputation score.
    pub async fn get_aggregate_score(&self, _agent_id: u64) -> Result<ReputationScore> {
        Err(Error::ContractError("Requires alloy provider setup".into()))
    }
}
